text/0000-build-std.md

```diff
@@ -276,52 +357,1098 @@ proposed by this RFC.
+  too, which checking for `#![no_std]` does not help with. Cargo could go
+  further and ask rustc whether a crate (or its dependencies) used `alloc`, but
+  this seems needlessly complicated.
+
+Furthermore, supporting explicit dependencies on standard library crates enables
+use of other Cargo features that apply to dependencies in a natural and
+intuitive way. If there were not explicit standard library dependencies and
+enabling features on the `std` crate was desirable, then a mechanism other than
+the standard syntax for this would be necessary, such as a flag (e.g.
+`-Zbuild-std-features`) or option in Cargo's configuration. This also applies to
+optional dependencies, public/private features, etc.
+
+**TODO:** document alternatives, or at least link to "why not automatic" below
+
+## Why disallow explicit builtin dependencies on other crates?
+[why-disallow-explicit-builtin-dependencies-on-other-crates]: #why-disallow-explicit-builtin-dependencies-on-other-crates
+
+## Why not add standard library dependencies to `Cargo.lock`?
+[why-not-add-standard-library-dependencies-to-cargo-lock]: #why-not-add-standard-library-dependencies-to-cargolock
+
+There is little advantage to adding standard library dependencies to
+`Cargo.lock`. As Cargo does not need to resolve the version of the standard
+library crates (it is always the version of the current toolchain), there would
+be nothing to record in the lockfile other than that the dependency exists,
+which isn't especially useful as the `Cargo.toml` already tracks this.
+
+## Why not migrate to always requiring explicit standard library dependencies?
+[why-not-migrate-to-always-requiring-explicit-standard-library-dependencies]: #why-not-migrate-to-always-requiring-explicit-standard-library-dependencies
+
+Explicit standard library dependencies with `builtin = true` will necessarily
+only be understood by newer versions of Cargo.
+
+If all packages were required to add explicit dependencies (perhaps over an
+edition or through some other mechanism), then every crate would require the
+newest version of Cargo to be understood, effectively raising the MSRV of every
+Rust crate.
+
+If only `no_std` crates (or crates with a `no_std` feature) add explicit
+dependencies on `core` or `alloc` then a much smaller percentage of the crates
+ecosystem will require the newest Cargo versions for their new explicit standard
+library dependencies to be understood.
+
+Alternative syntaxes, such as requiring `version = "*"` for explicit standard
+library dependencies, could be worthwhile to maintain a greater level of
+compatibility with older toolchain versions. However, as any currently accepted
+syntax would necessarily be interpreted differently by the build-std-supporting
+versions of Cargo, this doesn't really solve anything. For example, while
+`version = "*"` would be understood by older versions of Cargo, it would attempt
+to find the standard library crates on crates.io and fail. This is not a
```

Comment:
Maybe explore this a bit more? Eg, why don't we publish empty `std`, `core` and `alloc` crates there which resolves this issue? You then need a new Cargo to be able to use build-std but there's no ecosystem breakage by crates starting to adopt the explicit dependency.